### PR TITLE
Add code fix for MSTEST0017

### PR DESCRIPTION
**Pull request by engyebrahim**
fix: https://github.com/microsoft/testfx/issues/2947

### CODE REVIEW

File: src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        cancellationToken.ThrowIfCancellationRequested();
+
+        ArgumentListSyntax argumentList = invocationExpr.ArgumentList;
+        SeparatedSyntaxList<ArgumentSyntax> arguments = argumentList.Arguments;
+
+        ArgumentSyntax expectedArg = arguments.FirstOrDefault(arg => IsExpectedArgument(arg));
+        ArgumentSyntax actualArg = arguments.FirstOrDefault(arg => IsActualArgument(arg));
+
+        // Handle positional arguments if named arguments are not found
+        if (expectedArg == null || actualArg == null)
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs:63]: Is it working when only one of the 2 argument is using named argument?

File: src/Analyzers/MSTest.Analyzers.CodeFixes/CodeFixResources.resx [hunk 2]
```diff
@@ -123,4 +123,7 @@
   <data name="AssemblyInitializeShouldBeValidCodeFix" xml:space="preserve">
     <value>Fix signature</value>
   </data>
-</root>
\ No newline at end of file
+  <data name="FixAssertionArgsOrder" xml:space="preserve">
+    <value>Fix assertion args order</value>
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/CodeFixResources.resx]: I would go with either `Swap actual/expected arguments` or `Fix actual/expected arguments order`.

cc @nohwnd

File: src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        newArguments[actualIndex] = actualArg.WithExpression(tmpExpectedArg.Expression);
+
+        InvocationExpressionSyntax newInvocationExpr = invocationExpr.WithArgumentList(SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(newArguments)));
+        SyntaxNode root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+        SyntaxNode newRoot = root.ReplaceNode(invocationExpr, newInvocationExpr);
+
+        return document.WithSyntaxRoot(newRoot);
+    }
+
+    private static bool IsActualArgument(ArgumentSyntax argument) => string.Equals(argument.NameColon?.Name.Identifier.Text, "actual", StringComparison.Ordinal);
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs]: Please put the expression on the new line.

File: src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+    private static bool IsExpectedArgument(ArgumentSyntax argument) => string.Equals(argument.NameColon?.Name.Identifier.Text, "expected", StringComparison.Ordinal)
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs]: Please put the expression on the new line.

File: src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        var newArguments = arguments.ToList();
+        int expectedIndex = arguments.IndexOf(expectedArg);
+        int actualIndex = arguments.IndexOf(actualArg);
+
+        ArgumentSyntax tmpExpectedArg = expectedArg;
+        newArguments[expectedIndex] = expectedArg.WithExpression(actualArg.Expression);
+        newArguments[actualIndex] = actualArg.WithExpression(tmpExpectedArg.Expression);
+
+        InvocationExpressionSyntax newInvocationExpr = invocationExpr.WithArgumentList(SyntaxFactory.ArgumentList(SyntaxFactory.SeparatedList(newArguments)));
+        SyntaxNode root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs]: It's already computed on the lambda outside, it'd be better to just flow it to this method.

File: src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            expectedArg = arguments[0];
+            actualArg = arguments[1];
+        }
+
+        var newArguments = arguments.ToList();
+        int expectedIndex = arguments.IndexOf(expectedArg);
+        int actualIndex = arguments.IndexOf(actualArg);
+
+        ArgumentSyntax tmpExpectedArg = expectedArg;
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs]: This doesn't seem to be needed.

File: src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                context.Diagnostics);
+        }
+    }
+
+    private static async Task<Document> SwapArgumentsAsync(Document document, InvocationExpressionSyntax invocationExpr, CancellationToken cancellationToken)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        ArgumentListSyntax argumentList = invocationExpr.ArgumentList;
+        SeparatedSyntaxList<ArgumentSyntax> arguments = argumentList.Arguments;
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs]: Nit

```suggestion
        SeparatedSyntaxList<ArgumentSyntax> arguments = invocationExpr.ArgumentList.Arguments;
```

File: src/Analyzers/MSTest.Analyzers.CodeFixes/CodeFixResources.resx (hunk 2 above)
Comment (engyebrahim) [reply src/Analyzers/MSTest.Analyzers.CodeFixes/CodeFixResources.resx]: I used `Fix actual/expected arguments order` for now

File: src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs (hunk 1 above)
Comment (engyebrahim) [reply src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs:63]: yes and I added the two cases in the tests now

File: src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+        return document.WithSyntaxRoot(newRoot);
+    }
+
+    private static bool IsActualArgument(ArgumentSyntax argument) =>
+        string.Equals(argument.NameColon?.Name.Identifier.Text, "actual", StringComparison.Ordinal);
+
+    private static bool IsExpectedArgument(ArgumentSyntax argument) =>
+        string.Equals(argument.NameColon?.Name.Identifier.Text, "expected", StringComparison.Ordinal)
+                                                                       || string.Equals(argument.NameColon?.Name.Identifier.Text, "notExpected", StringComparison.Ordinal);
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/AssertionArgsShouldBePassedInCorrectOrderFixer.cs]:
```suggestion
        string.Equals(argument.NameColon?.Name.Identifier.Text, "expected", StringComparison.Ordinal)
        || string.Equals(argument.NameColon?.Name.Identifier.Text, "notExpected", StringComparison.Ordinal);
```

**Review by Evangelink — APPROVED** [review]

**Pull request merged**